File: leetcode/015-3sum/ThreeSum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> triplets = new LinkedList<>();
        Arrays.sort(nums);
        
        for (int i = 0; i < nums.length - 2 /* -2 = j + k which take +1 each */; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }
            
            int j = i + 1;
            int k = nums.length - 1;
            
            while (j < k) {
                int sum = nums[i] + nums[j] + nums[k];

                if (sum == 0) {
                    triplets.add(Arrays.asList(nums[i], nums[j], nums[k]));
                    
                    k--; j++;
                    while (j < k && nums[k] == nums[k + 1]) {
                        k--;
                    }
                    while (j < k && nums[j] == nums[j - 1]) {
                        j++;
                    }
                } else if (sum > 0) {
                    k--;
                } else {
                    j++;
                }
            }
        }
        
        return triplets;
    }
}

File: leetcode/015-3sum/three-sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& arr) {
      sort(arr.begin(), arr.end());

      vector<vector<int>> triplets;

      if (arr.size() < 3) {
        return triplets;
      }

      for (int i = 0; i < arr.size() - 2; i++) {
        if (i > 0 && arr[i] == arr[i - 1]) {
          continue;
        }

        int left = i + 1;
        int right = arr.size() - 1;

        while (left < right) {
          int sum = arr[i] + arr[left] + arr[right];
          if (sum == 0) {
            triplets.push_back({arr[i], arr[left], arr[right]});
            left++;
            right--;
            while (left < right && arr[left] == arr[left - 1]) { left++; }
            while (left < right && arr[right] == arr[right + 1]) { right--; }
          } else if (sum > 0) {
            right--;
          } else {
            left++;
          }
        }
      }

      return triplets;
    }
};
